bpf/process/types/probe_read_kernel_or_user.h

```diff
@@ -0,0 +1,150 @@
+ * BPF_CORE_READ_KERNEL_OR_USER(), in which final field is read into user-provided storage.
+ * See BPF_CORE_READ_KERNEL_OR_USER() below for more details on general usage.
+ */
+#define BPF_CORE_READ_KERNEL_OR_USER_INTO(userspace, dst, src, a, ...)         \
+	({                                                                     \
+		typeof(dst) dst_x = dst;                                       \
+		typeof(src) src_x = src;                                       \
+		typeof(a) a_x = a;                                             \
+		((userspace) ? (___core_read(bpf_core_read_user, dst_x,        \
+					     src_x, a_x, ##__VA_ARGS__))       \
+			     : (___core_read(bpf_core_read, dst_x, src_x, a_x, \
+					     ##__VA_ARGS__)))                  \
+	})
+
+/*
+ * BPF_CORE_READ_KERNEL_OR_USER_STR_INTO() does same "pointer chasing" as
+ * BPF_CORE_READ_KERNEL_OR_USER_STR() for intermediate pointers, but then executes (and returns
+ * corresponding error code) bpf_core_read_kernel_or_user_str() for final string read.
+ */
+#define BPF_CORE_READ_KERNEL_OR_USER_STR_INTO(userspace, dst, src, a, ...)                                       \
+	({                                                                                                       \
+		typeof(dst) dst_x = dst;                                                                         \
+		typeof(src) src_x = src;                                                                         \
+		typeof(a) a_x = a;                                                                               \
+		if (userspace)                                                                                   \
+			___core_read(bpf_core_read_user_str, dst_x, src_x,                                       \
+				     a_x, ##__VA_ARGS__) else ___core_read(bpf_core_read_str, dst_x, src_x, a_x, \
+									   ##__VA_ARGS__)                        \
+	})
+
+/*
+ * BPF_CORE_READ() is used to simplify BPF CO-RE relocatable read, especially
+ * when there are few pointer chasing steps.
+ * E.g., what in non-BPF world (or in BPF w/ BCC) would be something like:
+ *	int x = s->a.b.c->d.e->f->g;
+ * can be succinctly achieved using BPF_CORE_READ as:
+ *	int x = BPF_CORE_READ(s, a.b.c, d.e, f, g);
+ *
+ * BPF_CORE_READ will decompose above statement into 4 bpf_core_read (BPF
+ * CO-RE relocatable bpf_probe_read() wrapper) calls, logically equivalent to:
+ * 1. const void *__t = s->a.b.c;
+ * 2. __t = __t->d.e;
+ * 3. __t = __t->f;
+ * 4. return __t->g;
+ *
+ * Equivalence is logical, because there is a heavy type casting/preservation
+ * involved, as well as all the reads are happening through bpf_probe_read()
+ * calls using __builtin_preserve_access_index() to emit CO-RE relocations.
+ *
+ * N.B. Only up to 9 "field accessors" are supported, which should be more
+ * than enough for any practical purpose.
+ */
+#define BPF_CORE_READ_KERNEL_OR_USER(userspace, src, a, ...)              \
```

Comment:
Is this macro used somewhere?

Reply:
This is to support cases where people would otherwise use BPF_CORE_READ() or BPF_CORE_READ_USER() but where they only know programmatically which one they want. The idea of adding them at this point was so that they are consistent with the existing macros and encourage people to use them over alternatives.